### PR TITLE
Update DsUtils.cs

### PR DESCRIPTION
**Pull request by shell-with-it**
Updated getDomainSearchRoot to handle alternate UPN suffixes for domain. 

When Global Catalog is not used and the user upn contains a domain that is an alternate UPN suffix, SID lookup fails.

### CODE REVIEW

File: src/ADCS.SidExtension.PolicyModule/DsUtils.cs [hunk 1]
```diff
@@ -57,13 +57,24 @@ static class DsUtils {
 
         if (name.Contains("@")) {
             // upn
-            Int32 splitIndex = name.IndexOf("@", StringComparison.Ordinal);
-            domainTokens = name.Substring(splitIndex + 1).Split('.');
+            String domainPart = name.Substring(splitIndex + 1);
```
Comment (Crypt32) [inline src/ADCS.SidExtension.PolicyModule/DsUtils.cs:62]: `splitIndex` is not defined here. I believe that you accidentally removed a line where this variable is defined and initialized.

File: src/ADCS.SidExtension.PolicyModule/DsUtils.cs [hunk 2]
```diff
@@ -57,13 +57,24 @@ static class DsUtils {
...
+            using (var forest = Forest.GetCurrentForest()) {
+                using (var domain = forest.Domains.FirstOrDefault(d => d.Forest.Name.Equals(domainPart, StringComparison.OrdinalIgnoreCase))) {
```
Comment (Crypt32) [inline src/ADCS.SidExtension.PolicyModule/DsUtils.cs]: I'm not sure how it is supposed to work. All domains share same forest name, so `forest.Domain.FirstOrDefault` call makes little to no sense, because first domain entry will match the predicate and others are not evaluated. And there always will be one object -- forest root domain. In multi-domain forest it is going to fail. IIRC, domains in `forest.Domains` collection are ordered by name, so the predicate will match the first alphabetically ordered domain in the list. And it doesn't necessary have to be a domain where this particular UPN lives in.

In other words, I understand the problem, however this PR addresses the issue in a single-domain forest. Once you get two domains, it is going to fail.

In case of custom UPN suffix, it may be necessary to query all domains in a forest until match is found.

**Review by Crypt32 — CHANGES_REQUESTED** [review]

File: src/ADCS.SidExtension.PolicyModule/DsUtils.cs (hunk 1 above)
Comment (shell-with-it) [reply src/ADCS.SidExtension.PolicyModule/DsUtils.cs:62]: You're correct, your original definition should've remained. I have also updated it to iterate through each domain in the forest. I'm fairly new to C# so please forgive me.

**Review by shell-with-it — COMMENTED** [review]
Added ;

File: src/ADCS.SidExtension.PolicyModule/DsUtils.cs [hunk 3]
```diff
@@ -58,12 +58,23 @@ static class DsUtils {
...
             // upn
             Int32 splitIndex = name.IndexOf("@", StringComparison.Ordinal);
-            domainTokens = name.Substring(splitIndex + 1).Split('.');
+            //domainTokens = name.Substring(splitIndex + 1).Split('.');
+            String domainPart = name.Substring(splitIndex + 1);
+            using (var forest = Forest.GetCurrentForest())
+            {
+                foreach (Domain domain in forest.Domains)
+                {
+                    if (domain.GetDirectoryEntry().Properties["uPNSuffixes"].Contains(domainPart)) {
```
Comment (Crypt32) [inline src/ADCS.SidExtension.PolicyModule/DsUtils.cs:67]: I don't think that `uPNSuffixes` attribute is part of domain object. In fact, when you create a new UPN suffix, it is created for entire forest and stored in configuration naming context which isn't bound to any particular domain it is bound to forest root domain where configuration naming context is stored. In addition, this information can be cached in order to boost the performance. This will require additional code to support this. Let me think about this and I will address this issue. Anyways, thanks for raising this point and your efforts!

Comment (shell-with-it) [reply src/ADCS.SidExtension.PolicyModule/DsUtils.cs:67]: Thank you for the Policy module! Glad I could get this on your radar.
